Re-integrate RustPython parser repository

youknowone opened the pull request.

youknowone commented on crates/ruff/src/rules/flake8_bandit/helpers.rs:
```diff
@@ -10,10 +10,10 @@ static PASSWORD_CANDIDATE_REGEX: Lazy<Regex> = Lazy::new(|| {
     match &expr.node {
-        ExprKind::Constant {
+        ExprKind::Constant(ast::ExprConstant {
             value: Constant::Str(string),
             ..
-        } => Some(string),
+        }) => Some(string),
```
You may worry about this patterns by looking more verbose than before. Enum definition is changed to be using struct of enums. This is transitional state and it can be written like this later:
```suggestion
        ExprKind::Constant(constant) => Some(constant.value)
```

By allowing adding methods to `ast::ExprConstant` structs unlike before, it will be even better in the future.

MichaReiser replied:
I like it! The possibility for adding methods to `ExprConstant`, and that we can now pass specific `Stmt` instead of passing `Stmt` and than unwrapping a specific kind makes the API much more ergonomic

youknowone commented:
The failing tests seems not related and looks like github actions problem.

charliermarsh commented:
Oh very interesting, we now have separate structs for each variant. I'd love to get @MichaReiser's review.

youknowone commented:
Does udeps error related to this change? Not sure i have to fix it or not

youknowone commented:
The struct BestFitting is actually same named in 2 modules and used as public.
Renaming one seems better. Not sure why it is only triggered on this patch.

MichaReiser commented on crates/ruff/src/rules/flake8_2020/rules.rs:
```diff
@@ -175,21 +175,23 @@ pub fn subscript(checker: &mut Checker, value: &Expr, slice: &Expr) {
                     if let (
                         [Cmpop::Eq | Cmpop::NotEq],
-                        [Located {
+                        [Attributed {
```
I personally find `Attributed` somewhat more difficult to understand than `Located` because it's very generic. Attributed with what. But I don't feel strongly about it. 

A term that is used by many other compilers for a wrapper that adds source locations is [`Spanned`](https://www.google.com/search?client=firefox-b-d&q=rust+spanned) or (since we don't have a span type), could also name it `Ranged`. But again, I'm okay with keeping it. Just my two cents ;)

Note: I'm exploring how we can have location information on every node. We'll probably need this to correctly associate comments with nodes.

youknowone replied:
`Ranged` sounds good. Do you want it on this PR?

MichaReiser replied:
We can do it in another PR to avoid unnecessary rebasing

MichaReiser commented on Cargo.lock:
```diff
@@ -144,15 +144,6 @@ version = "1.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "d92bec98840b8f03a5ff5413de5293bfcd8bf96467cf5452609f939ec6f5de16"
 
-[[package]]
```
Oh nice, so many removed dependencies. This removes our total dependencies from 354 to 335!

MichaReiser commented on crates/ruff/src/rules/flake8_bugbear/rules/assignment_to_os_environ.rs:
```diff
@@ -20,13 +20,13 @@ pub fn assignment_to_os_environ(checker: &mut Checker, targets: &[Expr]) {
         return;
     };
-    if attr != "environ" {
+    if attr.as_str() != "environ" {
```
```suggestion
    if attr != "environ" {
```

This change isn't necessary because `Identifier` implements `PartialEq<str>`

MichaReiser commented on crates/ruff/src/rules/flake8_comprehensions/rules/unnecessary_list_call.rs:
```diff
@@ -45,7 +45,7 @@ pub fn unnecessary_list_call(checker: &mut Checker, expr: &Expr, func: &Expr, ar
         return;
     }
-    if !matches!(argument, ExprKind::ListComp { .. }) {
+    if !matches!(argument, ExprKind::ListComp(_)) {
```
Idea: It would be nice to have `argument.is_<node>()` methods on unions so that we could simplify this to 

```rust
if !argument.is_list_comp() {
	...
}
```

youknowone replied:
Yes, definitely. Actually that's also done in my local repository a few hours ago xD

MichaReiser commented on crates/ruff/src/rules/flake8_django/rules/model_without_dunder_str.rs:
```diff
@@ -95,24 +94,24 @@ fn checker_applies(checker: &Checker, bases: &[Expr], body: &[Stmt]) -> bool {
             continue
         };
-        if name != "Meta" {
+        if name.as_str() != "Meta" {
```
```suggestion
        if name  != "Meta" {
```

MichaReiser commented on crates/ruff/src/rules/pep8_naming/rules/invalid_first_argument_name_for_class_method.rs:
```diff
@@ -76,7 +76,7 @@ pub fn invalid_first_argument_name_for_class_method(
     }
     if let Some(arg) = args.posonlyargs.first().or_else(|| args.args.first()) {
-        if arg.node.arg != "cls" {
+        if arg.node.arg.as_str() != "cls" {
```
Would it make sense to also implement `PartialEq<str> for String` and `PartialEq<String> for &str` and (the same for `std::String`)?

MichaReiser reviewed (approved):
This is awesome. I was just thinking yesterday that I should look into how we can change our fork to fork the `Parser` repository instead and if we should introduce named variants for enums... and a few hours later, there's this PR! 

There are so many good things in here. I plan to merge this today, and we can discuss and solve some of the smaller nits later.

MichaReiser commented:
Should `String` and `Identifier` implement `Deref<str>`. It could simplify some of the code (e.g `option.as_deref()` would give you a `Option<&str>`).

youknowone commented:
I removed unnecessary `.as_str()` and made `Deref<str> for Identifier`. I wish I reverted them correctly.
@MichaReiser what should I do for udeps error?

MichaReiser commented:
Build-time: 
* Reduces the Total units for release builds from 345 to 309
* Reduces build time from 113 to 107s

Binary: 
* Increases by ~100Kb

Enum Sizes:
The size of `StmtKind` and `ExprKind` hasn't changed (They're still huge with 128 and 64 bytes, maybe we can reduce some of them by using `ThinVec`?)

MichaReiser commented:
> I removed unnecessary `.as_str()` and made `Deref<str> for Identifier`. I wish I reverted them correctly. @MichaReiser what should I do for udeps error?

Running `udeps` locally gives me:

```bash
cargo +nightly udeps
...
unused dependencies:
`ruff_python_ast v0.0.0 (/home/micha/astral/ruff/crates/ruff_python_ast)`
└─── dependencies
     └─── "ruff_rustpython"
`ruff_python_formatter v0.0.0 (/home/micha/astral/ruff/crates/ruff_python_formatter)`
└─── dependencies
     └─── "rustpython-common"
`ruff_rustpython v0.0.0 (/home/micha/astral/ruff/crates/ruff_rustpython)`
└─── dependencies
     ├─── "once_cell"
     └─── "rustpython-common"
Note: These dependencies might be used by other targets.
      To find dependencies that are not used by any target, enable `--all-targets`.
Note: They might be false-positive.
      For example, `cargo-udeps` cannot detect usage of crates that are only used in doc-tests.
      To ignore some dependencies, write `package.metadata.cargo-udeps.ignore` in Cargo.toml.
```

I'll look into it and push the changes so that we can then merge this right away

youknowone commented:
Ah, i had to see that parts, not the warning parts. That's because common is replaced by literal in a few places. I will do it.

youknowone commented:
Thank you for quickly merging it! It literally was making conflict for every a few hours 🤣

youknowone commented:
About the enum size, I think moving boxing point will be good. 

e.g. `StmtFunctionDef` wraps `args` with box. But StmtKind actually can box every variant.

```
#[derive(Clone, Debug, PartialEq)]
pub struct StmtFunctionDef<U = ()> {
    pub name: Identifier,
    pub args: Box<Arguments<U>>,
    pub body: Vec<Stmt<U>>,
    pub decorator_list: Vec<Expr<U>>,
    pub returns: Option<Box<Expr<U>>>,
    pub type_comment: Option<String>,
}
```

Does it make sense?

MichaReiser commented:
Boxing every variant is what https://github.com/Boshen/oxc/blob/main/crates/oxc_ast/src/ast/js.rs does

The downside is that pattern matching can become somewhat annoying without using nightly features (at least, that's what I remember to have heard). 

We may get away by only boxing the largest variants?

youknowone commented:
Do you have suggestions? The last option will be manually maintaining tables.
If we can do decide it from asdl data, it will be better.

currently `boxed` is related in https://github.com/RustPython/Parser/blob/main/ast/asdl_rs.py

MichaReiser commented:
> Do you have suggestions? The last option will be manually maintaining tables. If we can do decide it from asdl data, it will be better.
> 
> currently `boxed` is related in [RustPython/Parser@`main`/ast/asdl_rs.py](https://github.com/RustPython/Parser/blob/main/ast/asdl_rs.py?rgh-link-date=2023-05-11T08%3A22%3A54Z)

We can try boxing all of them to get a feel for the API's convenience. 

I'm considering to inline `range` into every node and add a `Ranged` trait instead that each node and all unions implement:

```rust
trait Ranged {
	fn range(&self) -> TextRange;

	fn start(&self) -> TextSize { self.range().start() }

	fn end(&self) -> TextSize { self.range().end() }
}
```

That removes the need for `Attributed` except for where someone wants to store a `custom` value with the node. For that use case I think it's best if a new struct is introduced per-use case (or we can keep `Attributed` around just for that use-case)

youknowone commented:
custom field also can be inlined to every node with `trait Custom`.
It will be easier to keep `Fold` working.
